### ISSUE TITLE
"Read Timeout!" Error When using RealSense D435i 

### ISSUE THREAD

**Issue by mbiason** [dumbotics/roboclaw_hardware_interface#13]
I've been using your hardware interface on my robot and it has been working great when using teleop with my RC controller. I am now moving onto SLAM and Nav2 for autonomous navigation. I picked up a D435i camera to begin with.

The "Read Timeout!" (from roboclaw_serial) happens when I run all the same nodes I do for teleop that work great, and when I plug in the camera to another USB port on my laptop, this error happens as soon as I move the joystick on my controller. I've even been able to drive in teleop with the camera unplugged, stop the rover, plug the camera in and as soon as I move the joystick again I get that error.

The roboclaw goes into a "frozen" state. There is no error led on the roboclaw itself but the two green leds are just constantly lit. Wheres in the normal state, the middle green pulses and top green lights up when commands are being executed.

I forgot to mention I use a RoboClaw 2x60AHV and do NOT have wheel encoders. I've also set up dev rules to no avail.

Any help would be appreciated. Thanks!

**Comment by mbiason** [comment]
I think I can confirm it is a communications issue between the computer (hardware serial interface?) and the correct device.

I was able to get everything correctly working and navigating to a goal pose on the actual rover by hooking up the roboclaw to the pi4 I also had on the robot and the camera and joystick to the laptop I was using. Running ros on both communicating via wifi, I was able to avoid the issue.

I wonder if this is an issue with the laptop and how it's usb interfaces are configured. It seems like when the hardware interface reads from the roboclaw it is not actually connecting to the right usb device, maybe the camera instead? Kinda weird.

**Comment by coxep** [comment]
Interesting... What happens if you plug in roboclaw first, start up the hardware interface, then plug in the realsense?

**Comment by mbiason** [comment]
I plug in the roboclaw and joystick USBs, start up the nodes for running in teleop like joy, hardware interface and rsp and drive the robot around in teleop mode just fine. If I leave everything running as is and only plug in the realsense, nothing happens immediately, but as soon as I move the joystick I get that error logged out and the roboclaw won't take anymore commands.

It's almost as if the `read` is reading from the wrong device to get position after the velocity is written. If I am understanding how it works, which I may not.

**Comment by coxep** [comment]
If you unplug the joystick while everything else is running just fine, do you get the error? Or does the error occur when you run everything except the joystick (restarting all nodes)?

**Comment by coxep** [comment]
I know you've looked into udev rules for roboclaw, but it really sounds like this might be the issue. Could you run this command when you're experiencing the issue?

`udevadm info -a -n /dev/roboclaw`

Replace `/dev/roboclaw` with your device.

**Comment by mbiason** [comment]
Here is that command output when ran after I see read timeout in the ros terminal.
[after.txt](https://github.com/user-attachments/files/17636575/after.txt)

**Comment by coxep** [comment]
Does dmesg show roboclaw disconnecting when the realsense is connected by any chance?

`sudo dmesg -T`

**Comment by mbiason** [comment]
It does not. Here is the output. For this I only had the roboclaw and realsense plugged in, no controller. Navigated in Rviz and got the read timeout when the rover started to move. It is actually a bit interesting because when doing it this way, the wheels keep moving constantly and won't stop until I cut power to the roboclaw. With teleop the wheels just flinch and don't move anymore.
[dmesg.txt](https://github.com/user-attachments/files/17684118/dmesg.txt)

**Comment by coxep** [comment]
I'd highly recommend setting a command timeout. You can do this in the BasicMicro windows application. I have my roboclaw set to stop motors after 0.1 seconds without getting any command.

I'm not sure what the issue is, but I'm guessing it's related to the realsense drawing a lot of power and this messing up the roboclaw connection. Have you tried connecting the realsense first, then the roboclaw?

Do you have a way to connect the roboclaw to the pi over the serial interface (gpio)?

**Comment by mbiason** [comment]
An interesting thing happened today. I disconnected both the motors to take the roboclaw inside to mess around with it. I hooked it up exactly like it is on the rover, with battery and everything including the realsense connected, I moved the joystick to teleop and I did not get the read timeout... The top roboclaw led pulsed like normal. 

I didn't confirm that the motor terminals were getting power with a meter like I should have but I assumed they were.

So I took it back outside to the rover hooked up the +/- of both motors, tired again and got the read timeout.

I am at a loss... Maybe it is a power issue, a ground issue, I have no idea.

I am actually not wanting to use the pi as I would like to have one device controlling everything. I was using a laptop to test everything out before I bought a mini pc to possibly use.

**Comment by mbiason** [comment]
I figured out the problem after thinking on it for a bit.

I had the camera mounted to the metal frame of the rover with a metal mount. The motors are also mounted to the frame via their metal mounts... So I removed the camera from the frame with everything else as is and voila! no errors!

There must be a ground issue or some kind of interference from the motors that when the motors got power it would go through the frame and camera and to the usb ports and mess up communication. I'll have to change the camera mount to a plastic one or something.

I appreciate you taking the time to comment on this tread and for making this hardware interface for the roboclaw!

- Matt
